### PR TITLE
Allow candidates to provide new referees

### PR DESCRIPTION
**Pull request by tijmenb**
## Context

This implements the functionality for replacing referees, when the referees the candidate provided haven't come back, their email address is bouncing, or they've refused to provide feedback.

## Changes proposed in this pull request

This implements the feature as described in the design:

https://bat-design-history.herokuapp.com/apply-for-teacher-training/add-a-new-referee
https://bat-design-history.herokuapp.com/apply-for-teacher-training/add-2-new-referees

It doesn't implement the "interstitial" page, that will happen in a follow up PR. 

## Guidance to review

Test it out locally.

## Link to Trello card

https://trello.com/c/oMo8ZCap

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: spec/system/candidate_interface/candidate_needs_to_provide_two_new_referees_spec.rb [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    @candidate = create(:candidate)
+    login_as(@candidate)
+  end
+
+  def and_i_have_submitted_my_application
+    @application_form = create(:completed_application_form, candidate: @candidate)
+  end
+
+  def and_both_of_my_referees_havent_responded_within_10_days
+    create(:reference, :requested, application_form: @application_form, created_at: Time.zone.now - 14.days)
```
Comment (stevehook) [inline spec/system/candidate_interface/candidate_needs_to_provide_two_new_referees_spec.rb]: 14 days versus 10 days?

File: app/components/candidate_interface/new_references_needed_component.html.erb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<div class="app-banner">
+  <div class="app-banner__message">
+    <% if reference_status.number_of_references_that_currently_need_replacing == 1 %>
+      You need to <%= govuk_link_to 'give details of a new referee', candidate_interface_new_additional_referee_path %>
+    <% else %>
```
[the remark this answers is not in the document]
Comment (tijmenb) [reply app/components/candidate_interface/new_references_needed_component.html.erb]: Yes! https://github.com/github/actionview-component/#conditional-rendering

File: app/views/candidate_interface/additional_referees/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<% content_for :title, title_with_error_prefix('Add a new referee', @reference.errors.any?) %>
+<% content_for :before_content, govuk_back_link_to(candidate_interface_application_form_path, 'Back') %>
```
Comment (tijmenb) [inline app/views/candidate_interface/additional_referees/new.html.erb]: Needs to sometimes go to the edit page of the previous reference.

File: app/views/candidate_interface/additional_referees/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+<% content_for :title, title_with_error_prefix('Add a new referee', @reference.errors.any?) %>
+<% content_for :before_content, govuk_back_link_to(candidate_interface_application_form_path, 'Back') %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <%= form_with model: @reference, url: candidate_interface_edit_additional_referee_path(@reference) do |f| %>
+      <%= f.govuk_error_summary %>
+
+      <h1 class="govuk-heading-xl">
+        Add a new referee
```
Comment (tijmenb) [inline app/views/candidate_interface/additional_referees/edit.html.erb]: Add the first new / second new referee.

File: app/views/candidate_interface/additional_referees/confirm.html.erb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            change_path: candidate_interface_edit_additional_referee_path(reference),
+          },
+          {
+            key: 'Relationship',
+            value: reference.relationship,
+            action: "relationship for #{reference.name}",
+            change_path: candidate_interface_edit_additional_referee_path(reference),
+          },
+        ] do %>
+        <%= render SummaryCardHeaderComponent, title: 'New referee' %>
```
Comment (tijmenb) [inline app/views/candidate_interface/additional_referees/confirm.html.erb]: Name as title.

File: app/components/candidate_interface/new_references_needed_component.html.erb (hunk 2 above)
[the remark this answers is not in the document]
Comment (stevehook) [reply app/components/candidate_interface/new_references_needed_component.html.erb]: I like that... know I see it!

File: app/services/reference_status.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+  def references_currently_in_draft
+    application_references.select(&:not_requested_yet?)
+  end
+
+  def references_that_need_to_be_replaced
+    refused = application_references.select(&:feedback_refused?)
+    bounced = application_references.select(&:email_bounced?)
+    ignored = application_references.select(&:feedback_requested?).select do |application_reference|
+      application_reference.created_at < TimeLimitConfig::DAYS_BEFORE_WE_ASK_FOR_A_NEW_REFEREE.days.ago
```
[the remark this answers is not in the document]
Comment (tijmenb) [reply app/services/reference_status.rb]: Updated to use the code added in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/1315.

File: app/services/reference_status.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    refused + bounced + ignored
+  end
+
+  def references_currently_requested
+    application_references.select(&:feedback_requested?).select do |application_reference|
+      application_reference.created_at > TimeLimitConfig::DAYS_BEFORE_WE_ASK_FOR_A_NEW_REFEREE.days.ago
```
[the remark this answers is not in the document]
Comment (tijmenb) [reply app/services/reference_status.rb]: Updated to use the code added in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/1315.

File: app/views/candidate_interface/additional_referees/confirm.html.erb [hunk 8]
```diff
@@ -0,0 +1,43 @@
+<% content_for :title, 'Check your referee’s details before confirming' %>
+<% content_for :before_content, govuk_back_link_to(candidate_interface_edit_additional_referee_path(@references.last), 'Back to application') %>
```
[the remark this answers is not in the document]
Comment (tijmenb) [reply app/views/candidate_interface/additional_referees/confirm.html.erb]: This is correct according to Paul, leaving it for now if that's okay.

File: spec/system/candidate_interface/candidate_needs_to_provide_two_new_referees_spec.rb (hunk 1 above)
Comment (tijmenb) [reply spec/system/candidate_interface/candidate_needs_to_provide_two_new_referees_spec.rb]: Updated (made it 30 days and updated the step definition)

File: app/services/reference_status.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  def references_currently_requested_and_not_overdue
+    application_references.select(&:feedback_requested?).reject(&:feedback_overdue?)
+  end
+
+  def feedback_completed?
+    application_references.select(&:feedback_provided?).size == 2
+  end
+
+  def application_references
+    @application_references ||= application_form.application_references.to_a
```
[the remark this answers is not in the document]
Comment (tijmenb) [reply app/services/reference_status.rb]: Yes it's for the `select`. Using the scopes will make separate database queries, while `to_a` and `select` will only ever use one query to get all references.

File: app/views/candidate_interface/additional_referees/confirm.html.erb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <h1 class="govuk-heading-xl">
+      Check your referee’s details before confirming
+    </h1>
+
+    <% @references.each_with_index do |reference, i| %>
```
[the remark this answers is not in the document]
Comment (tijmenb) [reply app/views/candidate_interface/additional_referees/confirm.html.erb]: Fixed.

File: app/services/reference_status.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    refused = application_references.select(&:feedback_refused?)
+    bounced = application_references.select(&:email_bounced?)
+    ignored = application_references.select(&:feedback_overdue?)
+
+    refused + bounced + ignored
+  end
+
+private
+
+  def references_currently_in_draft
```
[the remark this answers is not in the document]
Comment (tijmenb) [reply app/services/reference_status.rb]: Done!

File: app/components/candidate_interface/new_references_needed_component.html.erb [hunk 12]
```diff
@@ -0,0 +1,11 @@
+<div class="app-banner app-banner--info" aria-labelledby="info-message">
+  <div class="app-banner__message">
+    <h2 class="govuk-heading-m" id="info-message">
+      <% if reference_status.number_of_references_that_currently_need_replacing == 1 %>
+        You need to <%= govuk_link_to 'give details of a new referee', candidate_interface_new_additional_referee_path %>
```
Comment (tvararu) [inline app/components/candidate_interface/new_references_needed_component.html.erb]:
```suggestion
        You need to <%= govuk_link_to 'give details for a new referee', candidate_interface_new_additional_referee_path %>
```

Does this read better? IANACD

File: app/controllers/candidate_interface/additional_referees_controller.rb [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+      end
+    end
+
+  private
+
+    def find_reference
+      find_references.find(params[:application_reference_id])
+    end
+
+    def find_references
```
Comment (tvararu) [inline app/controllers/candidate_interface/additional_referees_controller.rb]:
```suggestion
    def not_requested_references
```

What do you think?

File: app/controllers/candidate_interface/additional_referees_controller.rb [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        redirect_to_confirm_or_show_another_reference_form
+      else
+        @reference = reference
+        render :edit
+      end
+    end
+
+  private
+
+    def find_reference
```
Comment (tvararu) [inline app/controllers/candidate_interface/additional_referees_controller.rb]:
```suggestion
    def current_reference
```

```suggestion
    def reference
```

What do you think of either of these names?

File: config/locales/application_form.yml [hunk 15]
```diff
@@ -304,6 +304,10 @@ en:
         - (should never appear)
         - First referee
         - Second referee
+        - Third referee
+        - Fourth referee
+        - Fifth referee
+        - Sixth referee
```
Comment (C-gyorfi) [inline config/locales/application_form.yml]: Is this logic assumes that a candidate would need to add a new request "a maximum of 6 times"?

**Comment by C-gyorfi** [comment]
I tried to recreate the behaviour locally with one declined reference:
image

When I log in as the candidate I did not get the prompt

File: config/locales/application_form.yml (hunk 15 above)
Comment (tijmenb) [reply config/locales/application_form.yml]: Yep! I'll make a note of this to improve it.

File: app/controllers/candidate_interface/additional_referees_controller.rb (hunk 13 above)
Comment (tijmenb) [reply app/controllers/candidate_interface/additional_referees_controller.rb]: Done

**Comment by tijmenb** [comment]
I've fixed the bug @C-gyorfi found, which required me to add a new column to the references to figure out whether or not they are "replacement" references.

**Pull request merged**